File: validate_bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void helper(TreeNode* root, vector<int> &ans, bool &check){
        if(root == nullptr){
            return;
        }
        helper(root->left, ans, check);
        if(ans.size() > 0){
            if(root->val <= ans[ans.size()-1]){
                check = false;
            }
        }
        ans.push_back(root->val);
        helper(root->right, ans, check);
    }
    bool isValidBST(TreeNode* root) {
        vector<int> ans;
        bool check = true;
        helper(root, ans, check);
        return check;
    }
};

File: validate_bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public void helper(TreeNode root, ArrayList<Integer> inorder){
        if(root == null) return;
        helper(root.left, inorder);
        inorder.add(root.val);
        helper(root.right, inorder);
    }
    public boolean isValidBST(TreeNode root) {
        ArrayList<Integer> inorder = new ArrayList<>();
        if(root == null) return true;
        helper(root, inorder);
        for(int i = 1 ; i <= inorder.size() - 1 ; i++){
            if(inorder.get(i-1) >= inorder.get(i)) return false;
        }
        return true;
    }
}